store/streaming/file/service.go

```diff
@@ -288,3 +178,31 @@ func isDirWriteable(dir string) error {
 	}
 	return os.Remove(f)
 }
+
+func writeLengthPrefixedFile(path string, data []byte, fsync bool) (err error) {
+	var f *os.File
+	f, err = os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err1 := f.Close(); err1 != nil && err == nil {
+			err = err1
+		}
+	}()
+	_, err = f.Write(sdk.Uint64ToBigEndian(uint64(len(data))))
+	if err != nil {
+		return err
+	}
+	_, err = f.Write(data)
+	if err != nil {
+		return err
+	}
+	if fsync {
+		err = f.Sync()
+		if err != nil {
+			return err
+		}
+	}
+	return
+}
```

Comment:
This code does a lot of work to accommodate partial error states which is probably not necessary.

```suggestion
func writeLengthPrefixedFile(path string, data []byte, fsync bool) error {
	f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
	if err != nil {
		return fmt.Errorf("open file: %w", err)
	}
	
	done := func() { f.Close() } // this is just a failsafe, so error is not relevant
	defer func() { done() }()
	
	if _, err := f.Write(sdk.Uint64ToBigEndian(uint64(len(data))); err != nil {
		return fmt.Errorf("write length prefix: %w", err)
	}
	
	if _, err := f.Write(data); err != nil {
		return fmt.Errorf("write data: %w", err)
	}
	
	if fsync {
		if err := f.Sync(); err != nil {
			return fmt.Errorf("sync: %w", err)
		}
	}
	
	done = func() {} // close is done explicitly now
	if err := f.Close(); err != nil {
		return fmt.Errorf("close: %w", err)
	}
	
	return nil
}
```

Reply:
The error wrapping looks good, but the close error handling, it seems to me the original one is better, it's handled in single place.

Reply:
BTW, [golang stdlib](https://cs.opensource.google/go/go/+/refs/tags/go1.19.3:src/os/file.go;l=722) use the same pattern.

Reply:
Close after an error doesn't require the error to be inspected, if the Close fails it doesn't matter, the overall operation failed anyway. In fact any error from Close should actually probably be ignored, as it's subsidiary to the original error. Whereas a Close after a series of successful writes does require the error to be inspected and returned, as if the Close fails it means the overall operation failed. Close has different meaning, and should be handled differently, in these two conditions, so handling it in a single block of code is I think no virtue.